lib/image.ml

```diff
@@ -228,35 +215,155 @@ let parse_src_with_size_info conf s =
+    | `Path p as path -> if Sys.file_exists p then Some path else None
     | `Empty -> (
         match full_portrait_path conf base p with
         | None -> None
-        | Some path -> Some (path, size_from_path path |> Result.to_option))
+        | Some path -> Some path)
   else None
 
-let get_portrait conf base p =
+(* In images/keydir we store either
+   - the old portrait as the original keydir.jpg/png/tif image
+   - the url to the portrait as content of a url.txt file
+*)
+let get_old_portrait conf base p =
+  let key = default_portrait_filename base p in
+  let p_dir =
+    String.concat Filename.dir_sep [ Util.base_path [ "images" ] conf.bname ]
+  in
+  let f = Filename.concat (Filename.concat p_dir "old") key in
+  (* TODO test for legal extensions *)
+  if Sys.file_exists (f ^ ".jpg") then Some (`Path (f ^ ".jpg"))
+  else if Sys.file_exists (f ^ ".jpeg") then Some (`Path (f ^ ".jpeg"))
+  else if Sys.file_exists (f ^ ".png") then Some (`Path (f ^ ".png"))
+  else if Sys.file_exists (f ^ ".gif") then Some (`Path (f ^ ".gif"))
+  else None
+
+let rename_portrait conf base p (nfn, nsn, noc) =
+  match get_portrait conf base p with
+  | Some (`Path old_f) -> (
+      let new_s = default_portrait_filename_of_key nfn nsn noc in
+      let old_s = default_portrait_filename base p in
+      let f = Filename.concat (Util.base_path [ "images" ] conf.bname) new_s in
+      let new_f = f ^ Filename.extension old_f in
+      (try Sys.rename old_f new_f
+       with Sys_error e ->
+         !GWPARAM.syslog `LOG_ERR
+           (Format.sprintf
+              "Error renaming portrait: old_path=%s new_path=%s : %s" old_f
+              new_f e));
+      let new_s_f =
+        String.concat Filename.dir_sep
+          [
+            Util.base_path [ "images" ] conf.bname;
+            "old";
+            new_s ^ Filename.extension old_f;
+          ]
+      in
+      let old_s_f =
+        String.concat Filename.dir_sep
+          [
+            Util.base_path [ "images" ] conf.bname;
+            "old";
+            old_s ^ Filename.extension old_f;
+          ]
+      in
+      (if Sys.file_exists old_s_f then
+       try Sys.rename old_s_f new_s_f
+       with Sys_error e ->
+         !GWPARAM.syslog `LOG_ERR
+           (Format.sprintf
+              "Error renaming old portrait: old_path=%s new_path=%s : %s" old_f
+              new_f e));
+      (* check for saved pictures in src/base/images/keydir *)
+      let old_d =
+        String.concat Filename.dir_sep
+          [ Util.base_path [ "src" ] conf.bname; "images"; old_s ]
+      in
+      let new_d =
+        String.concat Filename.dir_sep
+          [ Util.base_path [ "src" ] conf.bname; "images"; new_s ]
+      in
+      if Sys.file_exists old_d then
+        try Sys.rename old_d new_d
+        with Sys_error e ->
+          !GWPARAM.syslog `LOG_ERR
+            (Format.sprintf
+               "Error renaming portrait: old_path=%s new_path=%s : %s" old_d
+               new_d e))
```

Comment:
I'm not sure I understand, first we rename the portrait (a single image), and then we try to rename images that are in the images folder?

Reply:
do we need to move image notes too?

Reply:
At some point in time, I was saving portrait to images! 
Might be nice to have a fn.oc.sn.txt file associated with the portrait (same as for images in carrousel).